[LLDB][Telemetry]Define DebuggerTelemetryInfo and related methods

oontvoo opened the pull request:
- This type of entry is used to collect data about the debugger startup/exit
- Tests will be added (They may need to be shell test with a "test-only" TelemetryManager plugin defined. I'm trying to figure out how to get that linked only when tests are running and not to the LLDB binary all the time.

oontvoo committed:
[LLDB][Telemetry]Define DebuggerTelemetryInfo and related methods

- This type of entry is used to collect data about the debugger startup/exit
- Tests will be added (They may need to be shell test with a "test-only" TelemetryManager plugin defined. I'm trying to figure out how to get that linked only when tests are running and not to the LLDB binary all the time.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +60,83 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+
+struct DebuggerTelemetryInfo : public LLDBBaseTelemetryInfo {
+  std::string username;
+  std::string lldb_git_sha;
```
You may not have a git hash. How about storing the lldb version instead?

oontvoo replied:
Right - that's the general idea. But for our specific use-case, (as far as I remember) we need the git-sha to figure out which release it was (the info will be supplied at build time). I don't remember the exact detail, though.

will make this a generic "lldb_version" for now.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +60,83 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+  std::string description;
+};
+
+struct DebuggerTelemetryInfo : public LLDBBaseTelemetryInfo {
```
Looking at this event, I notice that it contains two kinds of information:

 - Information coming from lldb: the version and the exit status/description. 
 - Information coming from the environment lldb is run in: the username, the working directory and LLDB's path.

I think we should split this into two entries. For privacy reasons, I will never be interested in collecting the latter, but I'm very much interested in the former. If they're different entries, I can easily ignore on in my backend, but now I'd need to filter out individual fields.

oontvoo replied:
> I think we should split this into two entries.

There main reasons these were packed into one was  I/O : Fewer entries to send is generally better

>  For privacy reasons, I will never be interested in collecting the latter, but I'm very much interested in the former. If they're different entries, I can easily ignore on in my backend, but now I'd need to filter out individual fields.

Can you clarify why ignoring a particular type of entry is any more privacy compliant than ignoring some particular fields within an entry?

The principle here is that upstream LLDB collects as much relevant data as it can - then in each vendor-specific code, one can choose which fields to keep and which to ignore (You'd put such logic in your Destination class(es)).

labath replied:
I have many thoughts here. In no particular order, they are:
- I don't find the I/O argument particularly convincing, as I expect that any reasonable backend will want to cache/batch sending of these entries anyway
- I can believe that for analysis purposes, it's nicer to have a single "this is everything I know about the environment of the session" entry. I can also understand the desire for easy filtering of unwanted data. I don't know if there's a reasonable compromise here. Group this data into two sub-structs?
- the collect-then-filter approach makes sense for things which cannot be collected in any other way. However, pretty much everything being collected here is static data. There's no reason why it has to be collected at a particular location. (Like, even if we just stopped collecting the username here, a backend could just refetch it if it wanted to). Maybe we could not include things that recomputed in the backend? Or at least things that are likely to be problematic?

oontvoo replied:
I still don't understand how it would be much simpler filter out some structs. Then we would still have to make the decision which fields go into the the so-called "sensitive data struct". 

I imagine the vendor-specific code would look something like this

```
static DebuggerInfo* sanitizeFields(DebuggerInfo* entry) {
    // null out any fields you dont want to store
   // ....
}

Error VendorDestination::receiveEntry(DebuggerInfo* entry) {
    auto* newEntry = sanitizeFields(entry);
    storeToInternalStorage(entry);
}

```

And yes, it's more convenient to group all the context together (otherwise in downstream data structure, we'd need to do a bunch of joining to build a complete "picture").

JDevlieghere replied:
My arguments for separate events is that it's much easier to spot what's being collected: my backend is explicitly collecting events A, B and C and ignoring everything else. If I need to filter out fields that's a lot harder to audit. If you want to emit that as a single event in your backend, that's totally fine, but that's the responsibility of the back-end. I think it's much more important that the events are meaningful semantically.

oontvoo replied:
Ah, actually, we could make all the `atDebuggerStartup`,`atDebuggerExit`, etc methods virtual.

The upstream can collect a minimal set of common data.
Then the downstream's impl can override it to collect more (or less) data as needed.
Would that be an acceptable approach?

oontvoo replied:
P.S: updated the method to not collect any of the username, cwd, lldb_path (just leaving the fields empty).
(vendors can override the AtDebuggerStart() to collect those if they want to)

labath replied:
Okay, but the fields are still there. I know why you did this, but I think it's strange to have a field that is never set anywhere, and I definitely wouldn't want to define a field for every piece of information that anybody wants to collect. If we go in this direction (data collected downstream), then I think the fields should be defined downstream as well. One way to do that would be to wrap the upstream member in a vendor-specific event struct which includes the upstream struct as a member (or just a copy of its members). Another would be to attach them directly in the Destination class (that way the custom event struct doesn't even have to exist).

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -29,6 +30,9 @@ namespace telemetry {
 
 struct LLDBEntryKind : public ::llvm::telemetry::EntryKind {
   static const llvm::telemetry::KindType BaseInfo = 0b11000;
+  static const llvm::telemetry::KindType DebuggerInfo = 0b11001;
```
These values are kind of random. Any reason to not just one `1`, `2`, `3`, etc?

oontvoo replied:
Yes, these have common base so we can do a quick mask on all the subclasses's kind.
(See LLDBBaseTelemtryInfo::classof(), line 55):

```
 static bool classof(const llvm::telemetry::TelemetryInfo *t) {
    // Subclasses of this is also acceptable.
    return (t->getKind() & LLDBEntryKind::BaseInfo) == LLDBEntryKind::BaseInfo;
  }
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +60,83 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+
+  DebuggerTelemetryInfo() = default;
+
+  // Provide a copy ctor because we may need to make a copy before
+  // sanitizing the data.
+  // (The sanitization might differ between different Destination classes).
```
What's this sanitization. Does that exist yet? I thought the idea was that different emitters could decide what events to listen for? Did I miss that design change? It would be nice if we don't have to implement these trivial copy constructors.

oontvoo replied:
> What's this sanitization. 
This goes back to your previous privacy question.  It's where the Destination::receiveEntry() decides  whether to delete sensitive fields before sending the entry to some custom storage. If it does want to delete the fields, it'd do so by copying the original entry, then modifying the copy. Otherwise, it'd just send the original w/o copying,


> Does that exist yet?
No, it's up to the vendors to decide which fields to sanitize (because I bet we'd all disagree on this).

labath replied:
That makes sense to me, but it doesn't explain why we need to spell out the copying by hand. It's very easy to forget to update it (and it looks like even this implementation does not copy entries from the base class).

I thought this had something to do with this being being a polymorphic class (which usually have their copy constructors deleted -- for a reason), but AFAICT, all of the bases actually have their copy constructors (implicitly) defined. The possibility of object slicing is another complaint I have about this design (and I'm surprised it doesn't trigger any static analysis warnings).

In order to prevent object slicing, and avoid spelling out copy operations, how about we do something like [this](https://godbolt.org/z/GT591beeY). Basically, divide the two classes in the hierarchy into two kinds:
- abstract classes: these cannot be instantiated, so there's no danger of slicing. They still have a copy constructor, but they cannot be copied on their own.
- concrete final classes: these can be copied (using the default copy constructor implementation). However, they cannot be inherited from, which means they cannot slice anything.

oontvoo replied:
I've removed the copy ctor - agreed that it could be an impl detail. no need to spell it out here.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +60,83 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+  std::string target_uuid;
+
+  /// Set of key-value pairs for any optional (or impl-specific) data
+  std::map<std::string, std::string> meta_data;
```
```suggestion
  llvm::StringMap<std::string> meta_data;
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,6 +147,10 @@ class TelemetryManager : public llvm::telemetry::Manager {
+  // Each debugger is assigned a unique ID (session_id).
+  // All TelemetryInfo entries emitted for the same debugger instance
+  // will get the same session_id.
+  llvm::DenseMap<Debugger *, std::string> session_ids;
```
Should this store the debugger id instead of the debugger pointer? Debugger will get destroyed and this  seems like a foot-gun.

oontvoo replied:
When the debugger is destroyed, we'd remove it from the map. When does a Debugger object get destroyed? Is it when the process dies?

(Unless you're saying there could be multiple different Debugger objects for the same entity?  )

labath replied:
I think it could work, but I do believe its error-prone -- and maybe also unnecessary. I think we could avoid all this is we make the unique identifier of a debugger a pair. The first component would be the identifier of the TelemetryManager instance, generated when it's constructed (using random data, current time stamp, or whatever). The second component could the debugger ID (as retrieved by `debugger->GetID()`, which is a sequentially increasing integer). Then we don't need to store any per-debugger data,

The second component guarantees uniqueness of the identifier within a single process, the second does it "globally". If a particular backend doesn't like the fact that this is a pair, it can just concatenate them.

JDevlieghere commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -761,12 +767,29 @@ void Debugger::InstanceInitialize() {
 
 DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
                                     void *baton) {
+#ifdef LLVM_BUILD_TELEMETRY
+  lldb_private::telemetry::SteadyTimePoint start_time =
+      std::chrono::steady_clock::now();
+#endif
```
I think we should wrap this into an RAII object (like Progress) so that this does the right thing automatically.

oontvoo replied:
(NOT done - this is pending the result of whether we'd revert the ifdef approach and instead just make the cmake variable control llvm::telemetry::Config::EnableTelemetry)

JDevlieghere commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -761,12 +767,29 @@ void Debugger::InstanceInitialize() {
   }
   debugger_sp->InstanceInitialize();
+
+#ifdef LLVM_BUILD_TELEMETRY
+  if (auto *telemetry_manager = telemetry::TelemetryManager::getInstance()) {
+    if (telemetry_manager->getConfig()->EnableTelemetry) {
+      lldb_private::telemetry::DebuggerTelemetryInfo entry;
+      entry.start_time = start_time;
+      entry.end_time = std::chrono::steady_clock::now();
+      entry.debugger = debugger_sp.get();
+      telemetry_manager->atDebuggerStartup(&entry);
+    }
+  }
+#endif
```
I have to admit I didn't consider this implication when I added the CMake flag to make this configurable. I think we need to rethink this because checking the ifdef everywhere is going to get old really quickly.

oontvoo replied:
Can we convince the person that asked for this to accept the solution of setting `telemetry::Config::EnableTelemetry=false` and trusting that it'd do the right thign?

labath replied:
I'd very much like that, but I don't know how feasible is that. However, if that doesn't work out, I think we should still figure out a way to reduce the number of ifdefs. The RAII object might be way to do that (by defining a noop raii object if telemetry is disabled).

oontvoo replied:
> but I don't know how feasible is that.

I would really like to understand the objection. I recall the reasoning was something like it's faster to check for the non-existence of "Telemetry" symbol.  If that was the case, we can offer to define some statically defined symbol like DISABLE_TELEMETRY that would set the telemetry::Config::EnableTelemetry=false at build time and cannot be overriden.

I believe that would simplify a lot of things

JDevlieghere replied:
When I introduced the CMake variable, I had something in mind like what we do for signposts on macOS (`LLVM_SUPPORT_XCODE_SIGNPOSTS`) where we have a wrapper and then compile out support when it's not available. Admittedly, that's not what I did by conditionalizing the library, so that's on me and I'm happy to rectify that. 

That said, as I'm seeing how this is actually used in practice, I'm not so sure of the benefit of compiling things out, so maybe keeping the CMake variable and making it so that it's impossible to turn on at runtime if the flag is set, is good enough.

labath replied:
I would actually argue that the telemetry library (before the cmake flag) was very similar to the signposts setup (and all of the other cases where we have an external dependency), with the library *itself* being the wrapper: it exists, and can be used unconditionally, but unless some other component is also available, then it amounts to a slightly sophisticated noop.

Sure, in this case the subject matter is a lot more sensitive, and there are some differences in how it works. For example, that this "other component" takes the form of a subclass, and that no support for an actual implementation exists in this repo, but is rather meant to be added downstream. However, I think the principle is the same.

The last point (no actual concrete implementation) also makes me feel like any cmake flag would be mostly a placebo (as the only way to make this do something is to do it downstream, and at that point, you can always disable/override any protections we add here), but if that's what it takes, then that's fine. I think anything would be better than `#ifdef`ing everything out, or building another wrapper on top of that. I just don't know what kind of a form should that take. For example would something like
```
TelemetryManager* getInstance() {
#ifdef TELEMETRY
  return g_instance;
#else
  return nullptr;
#endif
}
```
be sufficient? (if we put that in a header, we should also have a decent chance of the compiler dead-stripping all the code which depends on this)

JDevlieghere commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1007,6 +1034,19 @@ void Debugger::Clear() {
     if (Diagnostics::Enabled())
       Diagnostics::Instance().RemoveCallback(m_diagnostics_callback_id);
+
+#ifdef LLVM_BUILD_TELEMETRY
+    if (auto telemetry_manager = telemetry::TelemetryManager::getInstance()) {
+      if (telemetry_manager->getConfig()->EnableTelemetry) {
+        // TBD: We *may* have to send off the log BEFORE the ClearIOHanders()?
+        lldb_private::telemetry::DebuggerTelemetryInfo entry;
+        entry.start_time = quit_start_time;
+        entry.end_time = std::chrono::steady_clock::now();
+        entry.debugger = this;
+        telemetry_manager->atDebuggerExit(&entry);
+      }
+    }
+#endif
```
This code is almost identical to the start time, this definitely an go into an RAII object that's shared between the two and calls the appropriate `atDebugger` method.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +60,83 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
 
+  const llvm::telemetry::Config *getConfig();
+
+  void atDebuggerStartup(DebuggerTelemetryInfo *entry);
+  void atDebuggerExit(DebuggerTelemetryInfo *entry);
```
It's a bit annoying since we're inheriting from LLVM, but we should use the LLDB coding style for methods we added, similar to `GetInstanceName` below.

oontvoo replied:
done

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -56,16 +54,107 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+  } else {
+    MiscTelemetryInfo misc_info;
+    misc_info.meta_data["internal_errors"] = "Cannot determine CWD";
+    if (auto er = dispatch(&misc_info)) {
+      LLDB_LOG(GetLog(LLDBLog::Object),
+               "Failed to dispatch misc-info at startup");
```
llvm::Error needs to be consumed (and it's better to make it's spell out its type, so that one can see that).

```suggestion
    if (llvm::Error er = dispatch(&misc_info)) {
      LLDB_LOG_ERROR(GetLog(LLDBLog::Object), std::move(er),
               "Failed to dispatch misc-info at startup: {0}");
```

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -56,16 +54,107 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+    }
+  }
+
+  if (auto er = dispatch(entry)) {
```
same here

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -56,16 +54,107 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+  // There must be a reference to the debugger at this point.
+  assert(entry->debugger != nullptr);
+
+  if (auto *selected_target =
+          entry->debugger->GetSelectedExecutionContext().GetTargetPtr()) {
+    if (!selected_target->IsDummyTarget()) {
+      const lldb::ProcessSP proc = selected_target->GetProcessSP();
+      if (proc == nullptr) {
+        // no process has been launched yet.
+        entry->exit_desc = {-1, "no process launched."};
+      } else {
+        entry->exit_desc = {proc->GetExitStatus(), ""};
+        if (const char *description = proc->GetExitDescription())
+          entry->exit_desc->description = std::string(description);
+      }
+    }
+  }
```
This is a fairly strange place to collect this kind of information. There's no guarantee that Process object will exist at the time the debugger terminates, or that it will have ran exactly once, etc.

If you're interesting in collecting the exit status, why not hook into a place that deals with that directly (somewhere around Process::SetExitStatus)?

oontvoo replied:
>There's no guarantee that Process object will  exist at the time the debugger terminates, or that it will have ran exactly once, etc.

Yes, that's why there' s a check - if there was no process when the debugger terminates (ie `proc == nullptr`), then we reports taht no process has been launched.



> If you're interesting in collecting the exit status, why not hook into a place that deals with that directly

This code is more interested in the exit of the *debugger* (ie., did the debugger exit normally or because some process crashed? or waht ...), which is why the telemetry-callback is done from Debugger's cleanup code.

labath replied:
> > There's no guarantee that Process object will  exist at the time the debugger terminates, or that it will have ran exactly once, etc.
> 
> Yes, that's why there' s a check - if there was no process when the debugger terminates (ie `proc == nullptr`), then we reports taht no process has been launched.

And if the process was launched twice this will get the last exit code -- is that what you want? What about the situation where the process is launched and then its target is deleted (`run; target delete; exit`)? Or when the process is launched, but then the user switches to debugging a different target (`run; target create/select; exit`)?

> 
> > If you're interesting in collecting the exit status, why not hook into a place that deals with that directly
> 
> This code is more interested in the exit of the _debugger_ (ie., did the debugger exit normally or because some process crashed? or waht ...), which is why the telemetry-callback is done from Debugger's cleanup code.

I don't see how one is relevant for the other. LLDB doesn't exit because the debugged process crashed. Any debuggee exit code is WAI as far as LLDB is concerned. I can see how you might be able to gather some very rough trends based on this information, but I think there may be better ways to achieve that (for example, I think that a better proxy for "was the user able to do something useful in this session" would be "whether the process exited without hitting any breakpoints". I also don't see why this would have to be gathered here, and I suspect its because you see lldb through the eyes of the DAP protocol, where there's a mostly one-to-one correspondence between a dap connection, a debugger object, and a debugged process. LLDB is much more general than that. It's true that a lot of command line users will use LLDB in a dap-like manner, but we actually want to discourage that as longer-lived sessions give you the benefit of all having all the information cached already.

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -56,16 +54,107 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+  if (!llvm::sys::fs::current_path(cwd)) {
+    entry->cwd = cwd.c_str();
+  } else {
+    MiscTelemetryInfo misc_info;
+    misc_info.meta_data["internal_errors"] = "Cannot determine CWD";
+    if (auto er = dispatch(&misc_info)) {
```
Is this actually useful? All it really does is tell you that determining CWD failed, but you could also find that out by looking at whether the `cwd` field is empty.

oontvoo committed:
Update lldb/source/Core/Telemetry.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

labath commented:
>  I'm trying to figure out how to get that linked only when tests are running and not to the LLDB binary all the time.

I don't think you can do that without building another lldb binary. We actually have something like that in the form of the [lldb-test binary](https://github.com/llvm/llvm-project/tree/main/lldb/tools/lldb-test). It's not an actual lldb, but it links all of the same components, and it shouldn't be a big problem to add a new component that's only used in lldb-test. I think it would be reasonable to add a new lldb-test sub-command which is designed to be used for testing the telemetry backend.

However, it's use tends to be somewhat cumbersome because you have to write some code do something, and then also drive that code from the command line. I would recommend trying to test as much as possible via the unit tests (which are basically can do anything that  lldb-test can, but you don't have to drive them from the command line), and use only when you really need an end-to-end experience.

oontvoo committed:
adddressed review comments:
 - remove collection of username,cwd,lldb_path (that can be collected in vendor-specific impl)
 - use DebuggerID as key rather than the debugger pointer
 - renaming

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -43,29 +77,79 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+      llvm::dyn_cast<LLDBBaseTelemetryInfo>(entry);
+  std::string session_id = m_id;
+  if (Debugger *debugger = lldb_entry->debugger) {
+    auto session_id_pos = session_ids.find(debugger->getID());
```
Can we drop the map? I doubt copying the string is noticably faster than constructing it each time (particularly if you use lower level constructs like `raw_string_ostream(lldb_entry->SessionId) << m_id << "_" << debugger->getID()`

Or even better, use two fields like I originally suggested? Then there's no copying and it has the advantage that you can choose how to do the analysis: I can imagine that in some situations you may want to see all events from a particular Debugger object, but in others you may want to see "everything that happens within a single process". I think that would be easier to do if the fields are separate.

oontvoo replied:
done

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -41,17 +41,32 @@ static uint64_t ToNanosec(const SteadyTimePoint Point) {
+// Generate a unique string. This should be unique across different runs.
+// We build such string by combining three parts:
+// <16 random bytes>_<timestamp>_<hash of username>
+// This reduces the chances of getting the same UUID, even when the same
```
Yes. done

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -56,16 +54,107 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+const Config *getConfig() { return m_config.get(); }
+
+void TelemetryManager::atDebuggerStartup(DebuggerTelemetryInfo *entry) {
+  UserIDResolver &resolver = lldb_private::HostInfo::GetUserIDResolver();
```
ok - will split this into two

oontvoo commented:
@JDevlieghere  @labath  : W.R.T the telemetry-build flag, I proposed [PR](https://github.com/llvm/llvm-project/pull/128534). Please have a look (ideally before continuing reviewing this patch since it'd simplify a few things here).
Thanks!

oontvoo committed:
remove unused fields

oontvoo committed:
Merge branch 'main' into debugger_info

oontvoo committed:
addressed review comments

oontvoo committed:
format

oontvoo commented:
I think this is ready for review now. New changes:
 - Incorporate the enable-telemetry flag from the other PR to get rid of all the ifdefs
 - remove process-exit/startup data from this - stick to debugger's exit/startup only. 
     + I'd like to find the right place to  report debugger's crashes (ie., in the code where it'd normally print "Please file a bug report ....")
 - add some RAII helper to reduce boilerplate

oontvoo committed:
try reporting the crash from a crash handler

oontvoo commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -226,13 +227,45 @@ lldb::SBError SBDebugger::InitializeWithErrorHandling() {
+
+static void InstallCrashTelemetryReporter() {
+#if LLVM_ENABLE_TELEMETRY
+
```
@labath Is this the right place to add the crash-handler?

labath replied:
Maybe? I don't think there's a single right place to do that, as each option comes with some tradeoffs. It may be best to leave this all up to downstream..

oontvoo replied:
agreed.

labath commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -226,13 +227,45 @@ lldb::SBError SBDebugger::InitializeWithErrorHandling() {
+#if ENABLE_BACKTRACES
+static void TelemetryHandler(void *) {
+  // TODO: get the bt into the msg?
+  // Also need to pre-alloc the memory for this entry?
+  lldb_private::telemetry::DebuggerInfo entry;
```
Another reason for a special entry point (or no entry point) is that you are using AtDebuggerExit in two fairly different ways. In `Debugger::Clear` you are calling it once per debugger, but if a crash happens (this function) you will just send one event (for all debuggers). If the crash happens sufficiently late in the shutdown process (after `Debugger::Clear` returns) you may even get both. Not the end of the world, but it kind of indicates that these are two different things.

oontvoo replied:
Ok, that's a great point. I'll revert this change and have the crash-handler registered in downstream code (with a new `AtCrash` callback to avoid confusion) Thanks!

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -38,29 +65,49 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
-  // attach the session_uuid to the entry.
+  // Assign the manager_id, and debugger_id, if available, to this entry.
+  LLDBBaseTelemetryInfo *lldb_entry =
+      llvm::dyn_cast<LLDBBaseTelemetryInfo>(entry);
```
```suggestion
      llvm::cast<LLDBBaseTelemetryInfo>(entry);
```

(as you're not checking (and I don't think you need to) whether the cast succeded)

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -38,29 +65,49 @@ void LLDBBaseTelemetryInfo::serialize(Serializer &serializer) const {
+}
+
+void TelemetryManager::AtDebuggerExit(DebuggerInfo *entry) {
+  // There must be a reference to the debugger at this point.
+  assert(entry->debugger != nullptr);
```
This isn't going to be true for crash events.

oontvoo replied:
as discussed in the other comment, the crash-handler will call a different method (e.,g atCrash())

labath commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -761,12 +764,27 @@ void Debugger::InstanceInitialize() {
+
+  if (helper.TelemetryEnabled()) {
+    helper.RunAtScopeExit([&]() {
+      lldb_private::telemetry::TelemetryManager *manager =
+          lldb_private::telemetry::TelemetryManager::GetInstance();
+      lldb_private::telemetry::DebuggerInfo entry;
+      entry.debugger = debugger_sp.get();
+      entry.start_time = helper.GetStartTime();
+      entry.end_time = helper.GetCurrentTime();
+      manager->AtDebuggerStartup(&entry);
```
This still feels rather boilerplate-y. `debugger`, `start_time` and `end_time` are present in every field, and declaring and sending an event is also something you need to do every time. What if the interface was something like:
```
ScopeTelemetryCollector<DebuggerInfo> telemetry_helper(debugger_sp.get());
```
where the object automatically fills declares the object, fills in the common fields and then sends it off in the destructor? It can provide some accessor to the object contained within it to let the code set any additional fields.

The only slightly variable part method that gets called to dispatch the event. But I see several ways to resolve that (roughly in order of preference):
1. ditch `AtDebuggerStartup` and have it always call `dispatch` (the customization can happen in `preDispatch`)
2. Declare a set of typed virtual methods with a single name (`virtual TypedDispatch(DebuggerInfo *); virtual TypedDispatch(OtherInfo*); ...`)
3. Have each entry type [declare the method that should get called](https://godbolt.org/z/vfjT6PxeE).
4. Have each call site declare the method that should get called: (`ScopeTelemetryCollector<DebuggerInfo> telemetry_helper(&TelemetryManager::AtDebuggerStartup, debugger_sp.get())`)

oontvoo replied:
> This still feels rather boilerplate-y. `debugger`, `start_time` and `end_time` are present in every field, and declaring and sending an event is also something you need to do every time. What if the interface was something like:
> 
> ```
> ScopeTelemetryCollector<DebuggerInfo> telemetry_helper(debugger_sp.get());
> ```
> 
> where the object automatically fills declares the object, fills in the common fields and then sends it off in the destructor? It can provide some accessor to the object contained within it to let the code set any additional fields.


How does the helper object know which fields to fill in?
For this specific case,  it's not much - just needed the debugger_sp. But for other cases (eg., commandinfo, it needs a bunch of local fields from callsite).
I'm worry we'll be adding too many layer of indirection. 



> 1. ditch `AtDebuggerStartup` and have it always call `dispatch` (the customization can happen in `preDispatch`)

These `AtDebuggerStartup`, etc, are how the downstream code can have overriding impl functions to collect additional data for each scenario.  The `preDispatch` is common for ALL scenarios (debugger-startup/shutdown/command-executation/target-load/etc).

JDevlieghere replied:
Pavel's suggestion is to fill in the common fields, i.e. the ones in `LLDBBaseTelemetryInfo`. I expressed my dislike for the `AtDebuggerStartup` and friends in another PR, and if the distinction is important, I'd argue that they should be different events. With that, you could do this: https://godbolt.org/z/eazvPWs46

oontvoo replied:
Ok - thx for the suggestion. I applied most of it with some modification (because we want to start the timer BEFORE the debugger_sp is created ).

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -7,12 +7,19 @@
 //===----------------------------------------------------------------------===//
 #include "lldb/Core/Telemetry.h"
 #include "lldb/Core/Debugger.h"
+#include "lldb/Core/Telemetry.h"
```
Are all of these actually used (in this patch)?

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +106,46 @@ class TelemetryManager : public llvm::telemetry::Manager {
+  }
+
+  bool TelemetryEnabled() {
+    TelemetryManager *instance = TelemetryManager::GetInstance();
+    return instance != nullptr && instance->GetConfig()->EnableTelemetry;
```
- If telemetry is disabled, the GetInstance() will return nullptr
- BUT if GetInstance() returns a valid object, telemetry might still be disabled (at runtime), hence the necessary check.

labath replied:
Okay, but why does GetInstance return a valid object if telemetry is disabled (at runtime)? Why does the TelemetryManager object even get created in this case? (Maybe there is a reason for that, but in that case, I'd like to understand why)

oontvoo replied:
In the original RFC, we proposed a way to toggle OFF/ON Telemetry (eg., via some simple command)

The idea is, the vendor can enable Telemetry at BUILD time. Then users can temporarily turn it off, then back on again. 

(It's probably already obviously but should be pointed out that the users can only turn OFF telemetry then ON again if Telemetry was already enabled to begin with.)

labath replied:
Thanks for the explanation. This makes sense, though I still think it's important to have a simple and succinct way to check whether telemetry is "actually" enabled at a given moment. Right now, we sort of have that as this detail is hidden in the RAII object, but if we find ourselves writing `manager && manage->IsEnabled()` often, would like to create helper function for just that (like a function that doesn't even let you access the telemetry manager object unless you're actually supposed to send something).

oontvoo commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -226,13 +227,45 @@ lldb::SBError SBDebugger::InitializeWithErrorHandling() {
 }
 
+#if LLVM_ENABLE_TELEMETRY
+#if ENABLE_BACKTRACES
```
Because to get the "exit-description" (ie., why did it crash), we need to print the stacktrace.
(Otherwise, i'm not sure there's a more descriptive way to describe the crash)

oontvoo committed:
revert crash handler changes

oontvoo commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -987,6 +1005,20 @@ void Debugger::Clear() {
+            lldb_private::telemetry::TelemetryManager::GetInstance();
+        lldb_private::telemetry::DebuggerInfo entry;
+        entry.debugger = this;
+        entry.exit_desc = {0, ""}; // If we are here, there was no error.
```
In the case of success, which is in this case, the exit_code is 0 and there's no description.
In the case of an error (which can only be caught in a crash-handler, as we discussed), the exit code and description can be something else.

This is NOT handling the exit of a process yet. That will be done in another PR (will be called from Process::SetExitDescription() )

labath replied:
Okay, but if we're not actually sending the crash reports upstream (which I think is the right choice), could we also avoid prescribing how a downstream implementation should represent those crashes. For example, someone might want to send the backtrace in a more structured form (list of addresses?) instead of a string. Fixed-size entries are a lot easier to handle in an async-signal context than a string (which basically cannot be used safely).

IOW, can we drop the `exit_desc` field from this struct?

oontvoo committed:
Update lldb/source/Core/Telemetry.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

oontvoo committed:
remove unused headers

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -21,14 +22,17 @@
 struct LLDBEntryKind : public ::llvm::telemetry::EntryKind {
   static const llvm::telemetry::KindType BaseInfo = 0b11000;
+  static const llvm::telemetry::KindType DebuggerInfo = 0b11001;
+  // There are other entries in between (added in separate PRs)
+  static const llvm::telemetry::KindType MiscInfo = 0b11110;
```
yes - no longer needed in this patch

oontvoo committed:
remove miscinfo kindtype

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -41,6 +45,7 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
   std::optional<SteadyTimePoint> end_time;
   // TBD: could add some memory stats here too?
 
+  uint64_t debugger_id = 0;
```
Yes - unless the ID could be zero? :thinking: 
I can make this an optional if ID could be zero?

labath replied:
Zero is probably fine (the first debugger instance seems to get `id=1`), but it's worth calling out explicitly.

oontvoo replied:
Ah, never mind. found it. `#define LLDB_INVALID_UID UINT64_MAX`
So I guess we could init this to `LLDB_INVALID_UID`

oontvoo committed:
Init debugger_id to LLDB_INVALID_UID

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +104,46 @@ class TelemetryManager : public llvm::telemetry::Manager {
+
+private:
+  SteadyTimePoint m_start;
+  std::stack<std::function<void()>> m_exit_funcs;
```
no, just wanted to communicate the fact that cleanup functions are executed in the reverse order in which they were created/queued .

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +104,46 @@ class TelemetryManager : public llvm::telemetry::Manager {
 };
 
+/// Helper RAII class for collecting telemetry.
+class ScopeTelemetryCollector {
```
We may need to run a few callbacks throughout (but not in this patch, so i'll apply the change there)

oontvoo committed:
addressed review comment

oontvoo committed:
Merge branch 'debugger_info' of github.com:oontvoo/llvm-project into debugger_info

oontvoo committed:
removed unused

oontvoo committed:
add smoke check

oontvoo committed:
remove unused headers

oontvoo committed:
added test

oontvoo committed:
formatting

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -41,6 +42,7 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
   std::optional<SteadyTimePoint> end_time;
   // TBD: could add some memory stats here too?
 
+  uint64_t debugger_id = LLDB_INVALID_UID;
```
```suggestion
  lldb::user_id_t debugger_id = LLDB_INVALID_UID;
```

or 

```suggestion
  std::optional<lldb::user_id_t> debugger_id;
```

JDevlieghere reviewed (commented):
A few more small comments but the overall approach LGTM.

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +100,50 @@ class TelemetryManager : public llvm::telemetry::Manager {
 private:
   std::unique_ptr<llvm::telemetry::Config> m_config;
+  // Each instance of a TelemetryManager is assigned a unique ID.
+  const std::string m_id;
```
What use case do you have in mind? Right now this field is read-only (copied onto multiple entry objects).

So it seems jsut storing a string directly seems more straightforward.

oontvoo replied:
> One reason for storing it in the UUID form might be if we think that someone could be interested in storing/transmitting the UUID in a binary form

This is the TelemetryManager - it is never transmitted or serialised anywhere. So why does compactness matter?

If you mean storing the id as UUID in the `TelemetryInfo` object, then we'd need to update the llvm base interface. However, the reason the id was a string in the base interface was because people wanted flexibility in what the id could be in each implementation.
Also the downside of storing it as UUID object is that you may have different `telemetry::Serializer` objects generating different values for the same UUID (ie.,different separators?). That's potentially problematic because ultimately, the point of this field is for grouping entries from the same session. We need the key to be identical across entries and not  "accidentally" differ because of serialisation methods.

labath replied:
> If you mean storing the id as UUID in the TelemetryInfo object, then we'd need to update the llvm base interface. However, the reason the id was a string in the base interface was because people wanted flexibility in what the id could be in each implementation.

I mean that, but yes, the fact that this is in the llvm base class make this tricky. I think this is fine then.

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +100,50 @@ class TelemetryManager : public llvm::telemetry::Manager {
 };
 
+/// Helper RAII class for collecting telemetry.
+template <typename I> struct ScopedDispatcher {
```
done - changed to Info.
(Isn't it usually `InfoT` ? not sure ...)

oontvoo committed:
change template to Info

oontvoo committed:
formatting

labath commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -761,7 +765,14 @@ void Debugger::InstanceInitialize() {
 
 DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
                                     void *baton) {
+  lldb_private::telemetry::ScopedDispatcher<
+      lldb_private::telemetry::DebuggerInfo>
+      helper([](lldb_private::telemetry::DebuggerInfo *entry) {
+        entry->lldb_version = lldb_private::GetVersion();
+      });
   DebuggerSP debugger_sp(new Debugger(log_callback, baton));
+  helper.SetDebugger(debugger_sp.get());
```
This is definitely better, but lambdas tend to add some boilerplate of their own. They may be useful in some cases (if you need to set some field at the very end) but I think that in many cases (this one included) we could just set the field directly, so I want to float this idea: I think this would be shorter if we had the dispatcher object expose the telemetry entry it is managing directly:
```suggestion
  telemetry::ScopedDispatcher<telemetry::DebuggerInfo> helper;
  DebuggerSP debugger_sp(new Debugger(log_callback, baton));
  if (telemetry::DebuggerInfo *info = helper.GetEntry()) {// returns NULL if telemetry is disabled
    info->debugger = debugger_sp.get();
    info->lldb_version = lldb_private::GetVersion();
  }
```

(and if we do need a callback, I don't see a reason why this couldn't coexist with that)

Any thoughts @JDevlieghere ?

oontvoo replied:
I would prefer not exposing the managed TelemetryInfo because in the next PR, i'm going to move it to being a local var in the execute function. (Reasons being I need to dispatch/execute more than one callbacks - one at the beginning and one at the end of the function. )

oontvoo replied:
On a second thought, how about we make a "NoOp" TelemetryManager that does not do anything in case when Telemetry is disabled? That way, we wouldn't need to check for the flag/ptr.

labath commented on lldb/unittests/Core/TelemetryTest.cpp:
```diff
@@ -95,3 +107,34 @@ TELEMETRY_TEST(TelemetryTest, PluginTest) {
+  lldb_private::FakePlugin::Initialize();
+  auto *ins = TelemetryManager::GetInstance();
+  ASSERT_NE(ins, nullptr);
+  std::vector<::llvm::telemetry::TelemetryInfo *> expected_entries;
```
```suggestion
  std::vector<::llvm::telemetry::TelemetryInfo *> received_entries;
```
(expected it the thing you expect to receive)

oontvoo replied:
done

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -56,13 +58,38 @@ struct LLDBBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+  std::string description;
+};
+
+struct DebuggerInfo : public LLDBBaseTelemetryInfo {
```
I definitely need this class to be extensible. 
Can you clarify what changes are needed to allow this?

labath replied:
`DebuggerInfo::classof(subclass_of_debugger_info)` should return true, which I guess means changing the implementation to something like `return T->getKind()&LLDBEntryKind::DebuggerInfo == LLDBEntryKind::DebuggerInfo`

oontvoo committed:
Update lldb/source/Core/Debugger.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

oontvoo commented:
@labath 
> IOW, can we drop the exit_desc field from this struct?

Well, we need a way to distinguish the start and the exit entries 
How about just keeping `exit_code` (which is set to zero)

> I still think it's important to have a simple and succinct way to check whether telemetry is "actually" enabled at a given moment

We can have a `static TelemetryManager::TelemetryCurrentEnabled()` which does the check.

labath commented:
> @labath
> 
> > IOW, can we drop the exit_desc field from this struct?
> 
> Well, we need a way to distinguish the start and the exit entries How about just keeping `exit_code` (which is set to zero)

You could distinguish them with a bool or enum field. I don't think an exit_code field makes sense if it's always going to be set to zero. I think it would make sense if you set it to `GetCommandInterpreter().GetQuitExitCode()`.

> 
> > I still think it's important to have a simple and succinct way to check whether telemetry is "actually" enabled at a given moment
> 
> We can have a `static TelemetryManager::TelemetryCurrentEnabled()` which does the check.

Maybe. Or maybe `TelemetryManager::GetInstanceIfEnabled` (not necessarily with that name) so you don't have to follow this up with a call to GetInstance(). Let's see how the code develops...

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -73,9 +100,50 @@ class TelemetryManager : public llvm::telemetry::Manager {
+template <typename Info> struct ScopedDispatcher {
+  // The debugger pointer is optional because we may not have a debugger yet.
+  // In that case, caller must set the debugger later.
+  ScopedDispatcher(std::function<void(Info *info)> callback,
```
done

oontvoo commented on lldb/unittests/Core/TelemetryTest.cpp:
```diff
@@ -18,25 +18,35 @@
-    received_entries->push_back(entry);
+    // Save a copy of the entry for later verification (because the original
+    // entry might have gone out of scope by the time verification is done.
+    if (auto *fake_entry = llvm::dyn_cast<FakeTelemetryInfo>(entry)) {
+      FakeTelemetryInfo *copied = new FakeTelemetryInfo();
+      copied->msg = fake_entry->msg;
+      copied->num = fake_entry->num;
+      received_entries->push_back(copied);
+    }
```
done

oontvoo committed:
address review comments

oontvoo committed:
Merge branch 'debugger_info' of github.com:oontvoo/llvm-project into debugger_info

oontvoo committed:
formatting

oontvoo commented:
Thanks, I'll merge this now. We can fine-tune the helper interface in the next PRs, which present slightly more complicated usecase.

oontvoo committed:
Merge branch 'main' into debugger_info
